// ngx-bar-rating/__ivy_ngcc__/rating.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { BarRatingComponent } from './component/bar-rating.component';
import { BarRatingPipe } from './pipe/bar-rating.pipe';
import * as ɵngcc0 from '@angular/core';
var BarRatingModule = (function () {
    function BarRatingModule() {
    }
BarRatingModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: BarRatingModule });
BarRatingModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function BarRatingModule_Factory(t) { return new (t || BarRatingModule)(); }, imports: [[
            CommonModule,
            FormsModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(BarRatingModule, { declarations: function () { return [BarRatingComponent,
        BarRatingPipe]; }, imports: function () { return [CommonModule,
        FormsModule]; }, exports: function () { return [BarRatingComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(BarRatingModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    BarRatingComponent,
                    BarRatingPipe
                ],
                imports: [
                    CommonModule,
                    FormsModule
                ],
                exports: [
                    BarRatingComponent
                ]
            }]
    }], function () { return []; }, null); })();
    return BarRatingModule;
}());
export { BarRatingModule };
/** @nocollapse */
BarRatingModule.ctorParameters = function () { return []; };
